File: src/hlp/api/routers/configurations.py
"""Configuration management router — admin only."""

from typing import Annotated

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from hlp.api.deps import get_db, require_admin
from hlp.api.schemas.configuration import (
    ConfigurationCreate,
    ConfigurationRead,
    ConfigurationUpdate,
)
from hlp.repositories import configuration_repository

router = APIRouter(prefix="/api/configurations", tags=["configurations"])


@router.get(
    "",
    response_model=list[ConfigurationRead],
    dependencies=[Depends(require_admin)],
)
def list_configurations(
    db: Annotated[Session, Depends(get_db)],
    type: str | None = Query(None, alias="type"),
    enabled: bool | None = None,
) -> list[ConfigurationRead]:
    items = configuration_repository.list_filtered(db, config_type=type, enabled=enabled)
    return [ConfigurationRead.from_model(c) for c in items]


@router.post(
    "",
    response_model=ConfigurationRead,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_admin)],
)
def create_configuration(
    db: Annotated[Session, Depends(get_db)],
    payload: ConfigurationCreate,
) -> ConfigurationRead:
    obj = configuration_repository.create(db, **payload.model_dump())
    db.commit()
    db.refresh(obj)
    return ConfigurationRead.from_model(obj)


@router.get(
    "/{config_id}",
    response_model=ConfigurationRead,
    dependencies=[Depends(require_admin)],
)
def get_configuration(
    db: Annotated[Session, Depends(get_db)],
    config_id: int,
) -> ConfigurationRead:
    obj = configuration_repository.get_or_raise(db, config_id)
    return ConfigurationRead.from_model(obj)


@router.patch(
    "/{config_id}",
    response_model=ConfigurationRead,
    dependencies=[Depends(require_admin)],
)
def update_configuration(
    db: Annotated[Session, Depends(get_db)],
    config_id: int,
    payload: ConfigurationUpdate,
) -> ConfigurationRead:
    update_data = payload.model_dump(exclude_unset=True)
    obj = configuration_repository.update(db, config_id, **update_data)
    db.commit()
    db.refresh(obj)
    return ConfigurationRead.from_model(obj)


@router.delete(
    "/{config_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_admin)],
)
def delete_configuration(
    db: Annotated[Session, Depends(get_db)],
    config_id: int,
) -> None:
    configuration_repository.delete(db, config_id)
    db.commit()


@router.post(
    "/{config_id}/toggle",
    response_model=ConfigurationRead,
    dependencies=[Depends(require_admin)],
)
def toggle_configuration(
    db: Annotated[Session, Depends(get_db)],
    config_id: int,
) -> ConfigurationRead:
    obj = configuration_repository.toggle_enabled(db, config_id)
    db.commit()
    db.refresh(obj)
    return ConfigurationRead.from_model(obj)

File: frontend/src/api/configurations.ts
import { api } from './client'

export interface ConfigurationRead {
  config_id: number
  config_type: string
  estate_id: number | null
  label: string
  url_or_path: string
  credentials_ref: string | null
  run_schedule: string | null
  enabled: boolean
  priority_rank: number
  notes: string | null
  scraping_config: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface ConfigurationCreate {
  config_type: string
  estate_id?: number | null
  label: string
  url_or_path: string
  credentials_ref?: string | null
  run_schedule?: string | null
  enabled?: boolean
  priority_rank?: number
  notes?: string | null
  scraping_config?: Record<string, unknown>
}

export interface ConfigurationUpdate {
  config_type?: string
  estate_id?: number | null
  label?: string
  url_or_path?: string
  credentials_ref?: string | null
  run_schedule?: string | null
  enabled?: boolean
  priority_rank?: number
  notes?: string | null
  scraping_config?: Record<string, unknown>
}

export async function listConfigurations(params?: {
  type?: string
  enabled?: boolean
}): Promise<ConfigurationRead[]> {
  const { data } = await api.get<ConfigurationRead[]>('/api/configurations', { params })
  return data
}

export async function getConfiguration(id: number): Promise<ConfigurationRead> {
  const { data } = await api.get<ConfigurationRead>(`/api/configurations/${id}`)
  return data
}

export async function createConfiguration(
  input: ConfigurationCreate,
): Promise<ConfigurationRead> {
  const { data } = await api.post<ConfigurationRead>('/api/configurations', input)
  return data
}

export async function updateConfiguration(
  id: number,
  input: ConfigurationUpdate,
): Promise<ConfigurationRead> {
  const { data } = await api.patch<ConfigurationRead>(`/api/configurations/${id}`, input)
  return data
}

export async function deleteConfiguration(id: number): Promise<void> {
  await api.delete(`/api/configurations/${id}`)
}

export async function toggleConfiguration(id: number): Promise<ConfigurationRead> {
  const { data } = await api.post<ConfigurationRead>(`/api/configurations/${id}/toggle`)
  return data
}
